compiler/rustc_hir_analysis/src/hir_ty_lowering/errors.rs

```diff
@@ -1215,7 +1216,111 @@ pub fn prohibit_assoc_item_binding(
+            //     If it's the first binding the preceding span would be
+            //     that of the last arg
+            //  3. Using this information work out whether the span
+            //     to remove will start from the end of the preceding span,
+            //     end at the start of the next span or will simply be the
+            //     span encomassing everything within the generics brackets
+
+            let Some(binding_index) = bindings.iter().position(|b| b.hir_id == binding.hir_id)
+            else {
+                bug!("a type binding exists but its HIR ID not found in generics");
+            };
+
+            let preceding_span = if binding_index > 0 {
+                Some(bindings[binding_index - 1].span)
+            } else {
+                args.last().map(|a| a.span())
+            };
+
+            let next_span = if binding_index < bindings.len() - 1 {
+                Some(bindings[binding_index + 1].span)
+            } else {
+                None
+            };
+
+            let removal_span = match (preceding_span, next_span) {
+                (Some(prec), _) => binding_span.with_lo(prec.hi()),
+                (None, Some(next)) => binding_span.with_hi(next.lo()),
+                (None, None) => {
+                    let Some(generics_span) = segment.args().span_ext() else {
+                        bug!("a type binding exists but generic span is empty");
+                    };
+
+                    generics_span
+                }
+            };
+
+            // Now emit the suggestion
+            if let Ok(suggestion) = tcx.sess.source_map().span_to_snippet(removal_span) {
+                e.span_suggestion_verbose(
+                    removal_span,
+                    "consider removing this type binding",
+                    suggestion,
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        };
+
+        // Suggests replacing the type binding with normal
+        // type argument (i.e. replacing <T = A> with <T>)
+        let suggest_direct_use = |e: &mut Diag<'_>, sp: Span, is_ty: bool| {
+            if let Ok(snippet) = tcx.sess.source_map().span_to_snippet(sp) {
+                let ty_or_const = if is_ty { "generic" } else { "const" };
+                e.span_suggestion_verbose(
+                    binding.span,
+                    format!("to use `{snippet}` as a {ty_or_const} argument specify it directly"),
+                    snippet,
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        };
+
+        // Check if the type has a generic param with the
+        // same name as the assoc type name in type binding
+        let generics = tcx.generics_of(def_id);
+        let binding_ident_lower = binding.ident.as_str().to_lowercase();
+        let matching_param =
+            generics.params.iter().find(|p| p.name.as_str().to_lowercase() == binding_ident_lower);
```

Comment:
> I wonder if it actually makes sense to do all that

Sorry, just to confirm, do you mean to say we don't need similar/best matching or we don't need case-insensitive matching?

Reply:
Right, that was ambiguously worded. I think only exact matching makes sense unless you can convince me otherwise. In case you do convince me, we could go further and generalize this to match similar names (`find_best_match_for_name` does case-insensitive matching, too).

Reply:
Yes, let's go with exact matching. The extra effort is not worth it if the likelihood of it being useful is low.